core/application/bloodDonationWorkflow/DonorSearchService.ts

```diff
@@ -0,0 +1,181 @@
+  ): Promise<void> {
+    const { seekerId, requestPostId, createdAt } = donorRoutingAttributes
+
+    const donorSearchRecord = await donorSearchRepository.getItem(
+      `${DONOR_SEARCH_PK_PREFIX}#${seekerId}`,
+      `${DONOR_SEARCH_PK_PREFIX}#${createdAt}#${requestPostId}`
+    )
+
+    if (donorSearchRecord === null) {
+      await donorSearchRepository.create({
+        id: donorRoutingAttributes.requestPostId,
+        ...donorRoutingAttributes,
+        status: DonationStatus.PENDING,
+        retryCount: 0
+      })
+    }
+    const isDonationUpdateRequest = sourceQueueArn === process.env.DONOR_SEARCH_QUEUE_ARN
+    const hasDonationCompleted = donorSearchRecord !== null && donorSearchRecord.status === DonationStatus.COMPLETED
+    const bloodQuantityUpdated = donorSearchRecord !== null && donorRoutingAttributes.bloodQuantity > donorSearchRecord.bloodQuantity
+
+    if (hasDonationCompleted && !isDonationUpdateRequest) {
+      return
+    }
+
+    if (isDonationUpdateRequest && hasDonationCompleted && bloodQuantityUpdated) {
+      await donorSearchRepository.update({
+        ...donorRoutingAttributes,
+        id: requestPostId,
+        status: DonationStatus.PENDING,
+        retryCount: 0
+      })
+    }
+
+    const retryCount = donorSearchRecord?.retryCount ?? 0
+    const updatedRecord: Partial<DonorSearchDTO> = {
+      ...donorRoutingAttributes,
+      id: requestPostId,
+      retryCount: retryCount + 1
+    }
+
+    if (retryCount >= Number(process.env.MAX_RETRY_COUNT)) {
+      updatedRecord.status = DonationStatus.COMPLETED
+      await donorSearchRepository.update(updatedRecord)
+      return
+    }
+
+    await donorSearchRepository.update(updatedRecord)
+
+    const stepFunctionPayload: StepFunctionInput = {
+      seekerId,
+      requestPostId,
+      createdAt,
+      donationDateTime: donorRoutingAttributes.donationDateTime,
+      requestedBloodGroup: donorRoutingAttributes.requestedBloodGroup,
+      bloodQuantity: Number(donorRoutingAttributes.bloodQuantity),
+      urgencyLevel: donorRoutingAttributes.urgencyLevel,
+      geohash: donorRoutingAttributes.geohash,
+      seekerName: userProfile.name,
+      patientName: donorRoutingAttributes.patientName ?? '',
+      location: donorRoutingAttributes.location,
+      contactNumber: donorRoutingAttributes.contactNumber,
+      transportationInfo: donorRoutingAttributes.transportationInfo ?? '',
+      shortDescription: donorRoutingAttributes.shortDescription ?? '',
+      city: donorRoutingAttributes.city,
+      retryCount: retryCount + 1,
+      message: getBloodRequestMessage(
+        donorRoutingAttributes.urgencyLevel,
+        donorRoutingAttributes.requestedBloodGroup,
+        donorRoutingAttributes.shortDescription ?? ''
+      )
+    }
+
+    await stepFunctionModel.startExecution(
+      stepFunctionPayload,
+      `${requestPostId}-${donorRoutingAttributes.city}-(${donorRoutingAttributes.requestedBloodGroup})-${Math.floor(Date.now() / 1000)}`
```

Comment:
is the datetime really required? given step function contains that information already!
could we instead use something more useful?
I am not sure what though 😅 
maybe the seeker id?

Reply:
smart 💯